[Pigeon]: version 0.1.0 (objc and testing features)

gaaclarke opened the pull request:
- cleaned up objc generation
- added ability to generate mock handlers
- added pigeon.dart

gaaclarke commented on packages/pigeon/mock_handler_tester/test/message.dart:
```diff
@@ -0,0 +1,147 @@
+// Autogenerated from Pigeon (v0.1.0), do not edit directly.
```
This file gets regenerated when running  `run_tests.sh`.  I still don't have the time to invest in creating a mechanism to generate these during standard execution of the tests.

gaaclarke commented on packages/pigeon/lib/pigeon_lib.dart:
```diff
@@ -30,7 +30,11 @@ const List<String> _validTypes = <String>[
 /// Metadata to mark an API which will be implemented on the host platform.
```
done

gaaclarke commented on packages/pigeon/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+
+* Added pigeon.dart.
+* Fixed some Obj-C linter problems.
+* Added the ability to generate a mock handler in Dart.
```
There is merit to what you are saying but `mock` is the terminology used by channels, that's why I chose it (https://api.flutter.dev/flutter/services/MethodChannel/setMockMethodCallHandler.html)

gaaclarke commented on packages/pigeon/pigeons/message.dart:
```diff
@@ -15,7 +15,7 @@ class SearchReply {
 }
 
-@HostApi()
+@HostApi(mockDartHandler: 'MockApi')
```
Not exactly the same but close.  I added a comment at the top.

gaaclarke commented on packages/pigeon/e2e_tests/test_objc/lib/dartle.dart:
```diff
@@ -128,3 +128,20 @@ class Api {
+  SearchReply search(SearchRequest arg);
+}
+
+void MockApiSetup(MockApi api) {
```
"mock handler" is already used terminology for the same concept elsewhere: https://api.flutter.dev/flutter/services/MethodChannel/setMockMethodCallHandler.html

xster replied:
Ah I see where you're going. That thing is a mock though. It's taking a production class and swapping it out for an alternate one which records and stubs interactions. 

This doesn't replace a production class. It's used in parallel with the production class to is a separate "working" class that has a different behavior that the real one. 

It'd be like we had `client` and `server` and we want to test `client`. We can have a `mockClient` which is what we all want but since we can't do it easily for existing tests, we're using `client` with a `fakeServer`. But if we do, we can't make the `fakeServer` be named `mockClient`. That would be really confusing. 

We can call this TestHost[name of API] for instance. Since we're specifically doing the 'host' part of the class annotated by `HostApi`

gaaclarke replied:
I renamed this class to `TestHostApi`.

gaaclarke commented on packages/pigeon/e2e_tests/test_objc/lib/dartle.dart:
```diff
@@ -128,3 +128,20 @@ class Api {
+  SearchReply search(SearchRequest arg);
+}
+
+void MockApiSetup(MockApi api) {
```
Changed it to `MockApi.setup()`.

xster replied:
oh I like it. What about the production class?

gaaclarke replied:
Yep, same.

gaaclarke commented on packages/pigeon/e2e_tests/test_objc/lib/dartle.dart:
```diff
@@ -128,3 +128,20 @@ class Api {
+  SearchReply search(SearchRequest arg);
+}
+
+void MockApiSetup(MockApi api) {
```
Changed it to `MockApi.setup()`.

xster replied:
I like it

gaaclarke commented on packages/pigeon/lib/pigeon_lib.dart:
```diff
@@ -30,7 +30,11 @@ const List<String> _validTypes = <String>[
-  const HostApi();
+  const HostApi({this.mockDartHandler});
+
+  /// The name of the interface to generate to receive Pigeon messages locally
```
Done.

gaaclarke commented on packages/pigeon/lib/pigeon_lib.dart:
```diff
@@ -30,7 +30,11 @@ const List<String> _validTypes = <String>[
+
+  /// The name of the interface to generate to receive Pigeon messages locally
+  /// for testing.
+  final String mockDartHandler;
```
as noted above

xster replied:
I'd still call this dartHostTestHandler

gaaclarke commented:
@xster I think I addressed all the comments.  There may still be a question about `mockDartHandler` let me know if that's still the case and what you decided you'd like it to be.  It will be a bit of a pain to change it and want to make sure to get it right.

xster commented on packages/pigeon/lib/pigeon_lib.dart:
```diff
@@ -27,13 +27,35 @@ const List<String> _validTypes = <String>[
-  const HostApi();
+  const HostApi({this.mockDartHandler});
+
+  /// The name of an interface generated next to the [HostApi] class.  Implement
```
nit: here and elsewhere, there's a bunch of double spaces.

xster commented on packages/pigeon/mock_handler_tester/README.md:
```diff
@@ -0,0 +1,3 @@
+# mock_handler_tester
+
+A test bed for testing the code generated by `mockDartHandler`.
```
if you rename it, keep the name consistent here and below

gaaclarke commented:
Landing before build-ipa's task finish.  They seem to be fubar, saying stuff about upgrading xcode but showing no progress.  Those tests previously passed before my renaming of the field.
